crates/sui-core/src/checkpoints/checkpoint_executor/utils.rs

```diff
@@ -267,3 +486,111 @@ impl TPSEstimator {
         self.tps
     }
 }
+
+#[cfg(test)]
+mod test {
+    use rand::{thread_rng, Rng};
+    use std::collections::HashMap;
+    use sui_macros::sim_test;
+
+    use super::*;
+    use futures::future::join_all;
+    use parking_lot::Mutex;
+
+    #[tokio::test]
+    #[should_panic(expected = "cannot skip stages")]
+    async fn test_skip_pipeline_stages() {
+        let stages = PipelineStages::new(0, CheckpointExecutorMetrics::new_for_tests());
+        let mut handle = stages.handle(0).await;
+        handle
+            .finish_stage(PipelineStage::WaitForTransactions)
+            .await;
+    }
+
+    #[sim_test]
+    async fn test_pipeline_stages() {
+        let stages = PipelineStages::new(0, CheckpointExecutorMetrics::new_for_tests());
+
+        let output_by_stage = Arc::new(Mutex::new(HashMap::new()));
+        let output_by_order = Arc::new(Mutex::new(Vec::new()));
+
+        let mut tasks = Vec::new();
+
+        for seq in 0..30 {
+            let stages = stages.clone();
+            let output_by_stage = output_by_stage.clone();
+            let output_by_order = output_by_order.clone();
+            tasks.push(tokio::spawn(async move {
+                let mut handle = stages.handle(seq).await;
+                let mut val = 0;
+                let mut get_next_val = || {
+                    val += 1;
+                    val
+                };
+
+                macro_rules! finish_stage {
```

Comment:
This macro would just make it harder to type the pipeline stage name since you no longer get typeahead from the enum. It could just be a function?

Reply:
sure, fixing